src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs

```diff
@@ -0,0 +1,377 @@
+                }
+            }
+
+            return CombineEditsIfPossible(newGroupedSemanticTokens, indexToEditKinds);
+        }
+
+        private static SemanticTokensEdit[] CombineEditsIfPossible(
+            SemanticToken[] newGroupedSemanticTokens,
+            Dictionary<int, SemanticTokenEditKind> indexToEditKinds)
+        {
+            // This method combines the edits into the minimal possible edits.
+            // For example, if an index contains both an insertion and deletion, we can combine the two
+            // edits into one.
+            // We can also combine edits if we have consecutive edits of certain types:
+            // Delete->Delete, Insert->Insert, Update->Update, Update->Insert, and Update->Delete.
+            // Note for the Update->Insert and Update->Delete cases, any further edits we combine can
+            // only be an Insert or Delete, respectively.
+
+            using var _ = ArrayBuilder<LSP.SemanticTokensEdit>.GetInstance(out var semanticTokensEdits);
+
+            // We sort the edit kinds by index since we need to know what kind of edits surround
+            // a given index in order to potentially combine them into one edit.
+            var editIndexes = indexToEditKinds.Keys.ToArray();
+            Array.Sort(editIndexes);
+
+            for (var i = 0; i < editIndexes.Length; i++)
+            {
+                var initialEditIndex = editIndexes[i];
+                var initialEditKind = indexToEditKinds[initialEditIndex];
+
+                if (initialEditKind.HasFlag(SemanticTokenEditKind.Insert) && initialEditKind.HasFlag(SemanticTokenEditKind.Delete))
+                {
+                    i = AddInsertionDeletionEdit(
+                        newGroupedSemanticTokens, indexToEditKinds, semanticTokensEdits, editIndexes, i, initialEditIndex);
+                }
+                else if (initialEditKind.HasFlag(SemanticTokenEditKind.Insert))
+                {
+                    i = AddInsertionEdit(
+                        newGroupedSemanticTokens, indexToEditKinds, semanticTokensEdits, editIndexes, i, initialEditIndex);
+                }
+                else
+                {
+                    i = AddDeletionEdit(
+                        indexToEditKinds, semanticTokensEdits, editIndexes, i, initialEditIndex);
+                }
+            }
+
+            return semanticTokensEdits.ToArray();
+
+            static int AddInsertionDeletionEdit(
+                SemanticToken[] newGroupedSemanticTokens,
+                Dictionary<int, SemanticTokenEditKind> indexToEditKinds,
+                ArrayBuilder<SemanticTokensEdit> semanticTokensEdits,
+                int[] editIndexes,
+                int i,
+                int initialEditIndex)
+            {
+                var deleteCount = 5;
+                var _ = ArrayBuilder<int>.GetInstance(out var tokensToInsert);
+                tokensToInsert.AddRange(newGroupedSemanticTokens[initialEditIndex].ToArray());
```

Comment:
instead of effectively allocating each token into an arrray, instead, have a method on the token that says "please add yourself to this ArrayBuilder at thsi position.  then we're alloc free here.

Reply:
I don't know if I fully understand this - maybe the name of the `ToArray` method is possibly confusing, since it's actually a method we define? I renamed it to `ConvertToArray`. But we're only converting one overall token here.